cpp/include/cudf/io/readers.hpp

```diff
@@ -0,0 +1,519 @@
+ public:
+  /**
+   * @brief Constructor for a filepath to dataset.
+   *
+   * @param filepath Path to whole dataset
+   * @param options Settings for controlling reading behavior
+   * @param mr Optional resource to use for device memory allocation
+   */
+  explicit reader(
+      std::string filepath, reader_options const &options,
+      rmm::mr::device_memory_resource *mr = rmm::mr::get_default_resource());
+
+  /**
+   * @brief Constructor for a memory buffer to dataset.
+   *
+   * @param buffer Pointer to whole dataset
+   * @param length Host buffer size in bytes
+   * @param options Settings for controlling reading behavior
+   * @param mr Optional resource to use for device memory allocation
+   */
+  explicit reader(
+      const char *buffer, size_t length, reader_options const &options,
+      rmm::mr::device_memory_resource *mr = rmm::mr::get_default_resource());
+
+  /**
+   * @brief Constructor for an Arrow file to dataset.
+   *
+   * @param file Arrow file object of dataset
+   * @param options Settings for controlling reading behavior
+   * @param mr Optional resource to use for device memory allocation
+   */
+  explicit reader(
+      std::shared_ptr<arrow::io::RandomAccessFile> file,
+      reader_options const &options,
+      rmm::mr::device_memory_resource *mr = rmm::mr::get_default_resource());
+
+  /**
+   * @brief Destructor explicitly-declared to avoid inlined in header
+   */
+  ~reader();
+
+  /**
+   * @brief Reads the entire dataset.
+   *
+   * @param stream Optional stream to use for device memory alloc and kernels
+   *
+   * @return `table` The set of columns
+   */
+  table read_all(cudaStream_t stream = 0);
+
+  /**
+   * @brief Reads and returns a range of rows.
+   *
+   * @param skip_rows Number of rows to skip from the start
+   * @param num_rows Number of rows to read; use `0` for all remaining data
+   * @param stream Optional stream to use for device memory alloc and kernels
+   *
+   * @return `table` The set of columns
+   */
+  table read_rows(size_type skip_rows, size_type num_rows,
+                  cudaStream_t stream = 0);
+};
+
+}  // namespace avro
+
+//! CSV format
+namespace csv {
+
+/**
+ * @brief Options for the CSV reader.
+ */
+struct reader_options {
+  compression_type compression = compression_type::AUTO;
+  std::vector<std::string> columns;
+
+  std::vector<std::string> column_filter_by_name;
+  std::vector<int> column_filter_by_index;
+  size_type header_row = 0;
+
+  char terminator = '\n';
+  char delimiter = ',';
+  char thousands_char = '\0';
+  char decimal_char = '.';
+  char comment_char = '\0';
+
+  bool windowslinetermination = false;
+  bool whitespace_delimiter = false;
+  bool ignore_leading_whitespace = false;
+  bool ignore_empty_lines = true;
+
+  bool quoting = true;
+  char quote_char = '"';
+  bool double_quote = true;
+
+  std::unordered_map<std::string, std::shared_ptr<data_type>> column_types;
+  std::vector<std::string> true_values{"True", "TRUE", "true"};
```

Comment:
Consider making these a `constexpr std::array`.

Reply:
I think despite the `read_csv()` documentation, PANDAS has default sequences that it recognizes as boolean true/false. I wasn't sure whether to explicitly expose them here as part of the user-specified input, or hide them internally like the legacy code.

It's exposed externally so that I can avoid having extra flags to clear the defaults or not. The user can either append to the list, clear the list, or keep as default...